### PR TITLE
Game saving code extracted

### PR DESCRIPTION
**Pull request by uHappyLogic**
Lying foundation for new save game handling system.
Code that is responsible for game saving moved to separate units, which enables unit testing.
Game saving and loading checked.

### CODE REVIEW

File: src/GameSaves/OriginalGameSaveToFileWriter.cs [hunk 1]
```diff
@@ -777,7 +785,7 @@ public void WriteSaveGame(GameSave gameSave)
                     bw.Write((byte)0);
                 }
 
-                // TODO: AI opponents
+                // TODO: AI opponents (shoulnd't be - 1?)
```
Comment (SWY1985) [inline src/GameSaves/OriginalGameSaveToFileWriter.cs]: It should be -2, because Barbarians are counted as a player, and both the Human player and the Barbarians are not handled as opponents in the original game.

File: src/GameSaves/OriginalGameSaveToFileWriter.cs [hunk 2]
```diff
@@ -777,7 +785,7 @@ public void WriteSaveGame(GameSave gameSave)
...
                 bw.Write((ushort)(gameSave.Players.Count - 2));
```
Comment (SWY1985) [inline src/GameSaves/OriginalGameSaveToFileWriter.cs]: Perhaps a better way to write this is something like:
`bw.Write((ushort)(gameSave.Player.Count(x => !(x.Civilization is Barbarian) && !x.IsHuman)));`

File: src/Player.cs [hunk 3]
```diff
@@ -316,11 +316,11 @@ public string LatestAdvance
 			}
 		}
 
-		public IAdvance[] Advances
+		public List<IAdvance> Advances
```
Comment (SWY1985) [inline src/Player.cs]: Why did you change the array to a List?

File: src/Screens/SaveGame.cs [hunk 4]
```diff
@@ -204,7 +207,10 @@ public override bool KeyDown(KeyboardEventArgs args)
 				if (_gameId >= 0)
 				{
 					SaveGameFile file = GetSaveGames().ToArray()[_gameId];
-					Game.Save(file.SveFile, file.MapFile);
+
+                    var saver = new OriginalGameSaveToFileWriter(file.SveFile, file.MapFile);
```
Comment (SWY1985) [inline src/Screens/SaveGame.cs]: Please use the strong type `OriginalGameSaveToFileWriter` instead of `var`

File: src/GameSaves/OriginalGameSaveToFileWriter.cs [hunk 5]
```diff
@@ -35,11 +35,11 @@ public void WriteSaveGame(GameSave gameSave)
             {
                 ushort randomSeed = Map.Instance.SaveMap(MapFile);
                 ushort activeCivilizations = 1;
-                for (int i = 1; i < gameSave.Players.Count; i++)
+                for (int i = 1; i < gameSave.Players.Count; ++i)
```
Comment (SWY1985) [inline src/GameSaves/OriginalGameSaveToFileWriter.cs]: Using ++i instead of i++ doesn't change anything, as far as I know?

File: src/Units/UnitsCreator.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+// You should have received a copy of the CC0 legalcode along with this
+// work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+
+
+using CivOne.Enums;
+using CivOne.Interfaces;
+
+namespace CivOne.Units
+{
+    public class UnitsCreator
```
Comment (SWY1985) [inline src/Units/UnitsCreator.cs]: By making the class static, you won't be able to construct an empty class, like so:
`public static UnitsCreator`

**Comment by SWY1985** [comment]
More than a month without activity. Probably no longer compatible with the current core restructuring.

File: src/Screens/SaveGame.cs (hunk 4 above)
Comment (jakebrown58) [inline src/Screens/SaveGame.cs]: SWY1985 - 'var' is a strong type in C# - the difference is that it implicitly vs explicitly typed.  It will compile to the same code if you use 'var' or the explicit name ('OriginalGameSavetoFileWriter) in this case.  https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/var

...if it's a style choice though - fair enough, since it's your codebase and your style guide.

Comment (SWY1985) [inline src/Screens/SaveGame.cs]: I know it's not wrong. It's purely a style choice.
